llvmlite/ir/types.py

```diff
@@ -109,30 +112,67 @@ def _to_string(self):
 class PointerType(Type):
     """
     The type of all pointer values.
+    By default (without specialisation) represents an opaque pointer.
     """
+    is_opaque = True
     is_pointer = True
     null = 'null'
 
-    def __init__(self, pointee, addrspace=0):
-        assert not isinstance(pointee, VoidType)
-        self.pointee = pointee
+    # Factory to create typed or opaque pointers based on `pointee'.
+    def __new__(cls, pointee=None, addrspace=0):
+        if cls is PointerType and pointee is not None:
+            return _TypedPointerType(pointee, addrspace)
```

Comment:
I made a mistake in my suggestion below for the double-`__init__` issue - I think it should have been this:
```suggestion
            return super().__new__(_TypedPointerType)
```

Reply:
Thanks very much for noting this - I've updated it!